### ISSUE TITLE
webui: sessions are not destroyed when user logs out

### ISSUE THREAD

**Issue by engerm** [RagtagOpen/nomad#228]
Sessions are not destroyed when a user logs out. A session cookie after a user logs out is still valid.

Steps:

1) log in as a user

2) Create a car

3) On the carpool landing page, capture the URL an get the session and remember_token (or capture the GET request in burp

4) in the browser, log out (can even close out the browser and clear cache)

5) Replay the same request using the same session and remember_token

**Actual Results:**

the requests still go through and are able to access the User only page that allows to "Cancel your carpool" and Approve/Deny Ride Requests.

**Expected Results:**

1) Sessions would be destroyed when a user logs out

2) Would not be able to use sessions after a user have logged out

The below request is from the ELT account

```
GET /carpools/26e52077-6c1f-40ea-81ba-0a8344e9a551 HTTP/1.1
Host: nomad.ragtag.org
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:55.0) Gecko/20100101 Firefox/55.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: https://nomad.ragtag.org/carpools/mine
Cookie: session=.eJwdj0FuQyEMRK9Ssc7CGGMgl4nAHqtVpVb6P1lFuXto9_PmzTzTLQ6cn-l6Px64pNuXp2siAHlGjBrBNDxz7wpxj8zcnTuNPOoQsE7y4lbmTnVqXIxRCkSYfVSbk5rmEKpF-8rSbZHX2VomJWkROYhXYKENNkOfGzNJl2TnEbf77zd-9h4bmLui-qZb0zFoO7jUKtTrAi3nuaC6uceJ4__EM308_q4U5YYm28OStz8yAkZY7gq19Hq9AZvWSdA.DLm78w.W72sRuphghxx9kS6PswaUKOKn0A; remember_token=03627e74-f022-418c-bf1e-fec0ebdd6e6c|98172265deaf71f9415c394346b4ecf6018d84a6c6a9da8afde8da7148aa9a9952c74484f43bdb47e5b2850fcd68048c1c924a2bef4fd9888770f0ab743a6583
DNT: 1
Connection: close
Upgrade-Insecure-Requests: 1
```

**Comment by iandees** [comment]
I think all of these cookie-related issues can be solved by enabling strict session control in flask-login as described here: https://flask-login.readthedocs.io/en/latest/#cookie-settings

**Comment by simonl2002** [comment]
This issue still exists. I created a carpool captured my cookies and then logged out.
Curling the carpool from the command line with my old cookies, I am still able to see the logged in version of the page.

We need to actually delete the session from the DB once the /logout request is made.

**Comment by iandees** [comment]
No sessions are stored in the database.

**Comment by simonl2002** [comment]
OK so talked with @iandees about this. Because we don't have server side storage of sessions there is currently no way to invalidate them. The current scheme is just checking whether the signed cookie is valid. 

Option would be to use [flask-session](https://pythonhosted.org/Flask-Session/) which is recommended in this post.
https://blog.miguelgrinberg.com/post/how-secure-is-the-flask-user-session

Additionally we should assess whether we should even store any information related to the user in the cookie at all.

**Comment by iandees** [comment]
The only user-related information in the session is [the "remember me" cookie](https://flask-login.readthedocs.io/en/latest/#remember-me) storing the user's UUID. We could try ditching that cookie and see what happens?

**Comment by joearasin** [comment]
Strict sessions don't really have much to do this -- they affect what happens when a session that is set from browser/source is attempted to be used in another browser.

In general, the cookie is the point of truth for a session -- It is a signed bundle of data that's stored client-side. The server will know if the client is trying to send it data that it didn't send the other way, but "the browser forgetting the session" means "the session is no more" (from a practical standpoint).

**Comment by sjohnson540** [comment]
It makes sense to store the sessions in a database. I have used Redis to store sessions for Flask apps in the past, and that seems like would be the best path going forward. Flask-login doesn't really provide much in the way of security it is just some sugar on top of storing sessions directly in Flask's default session storage. I am happy to add the Redis session storage if everyone agrees it is a good idea.

**Comment by iandees** [comment]
Can we just tell the browser to ditch the session cookie when the user clicks logout?

**Comment by joearasin** [comment]
@iandees I think that’s exactly what happens right now (or at least it should). 

Re: Storing sessions server-side... Do we need to? What is the particular thing we’re concerned about?

**Comment by joearasin** [comment]
That is to say — it’s another piece of infrastructure to deploy/maintain... What problem are we having now that it solves?

**Comment by iandees** [comment]
The original issue here was that user sessions aren't destroyed on logout. The only way we store session information is on the cookie. If the cookie is destroyed on login, then this issue can be closed with no change 😄 .

**Comment by iandees** [comment]
I think the suggestion for using redis or database came in because then we could proactively remove session information from that session store on logout by deleting the entry for that user.

**Comment by sjohnson540** [comment]
The problem is that you don't really have control over the user's session. If they want to hold onto the cookie, or if the session information is stolen, then someone could impersonate that user fairly easily. In the case of an Admin user, that becomes extra harmful.

If that's not something you all are worried about, then that's fine, but its definitely a security hole to be aware of.

**Comment by simonl2002** [comment]
@sjohnson540 is correct. If a user's session info is stolen there's nothing even the service admins could do short of changing the secret. If we don't want to add the need for another piece of infrastructure storing them in the database would work fine.

**Comment by joearasin** [comment]
Re: Nothing short of changing the secret -- not exactly. The [alternative token](https://flask-login.readthedocs.io/en/latest/#alternative-tokens) pattern in the flask-login docs provides a way to invalidate all sessions for one user in a particularly straightforward manner, so on password change or user permissions change this could be implemented.

When would admins need to invalidate one particular user session w/o invalidating all sessions for a particular user?



Implementin

**Comment by sjohnson540** [comment]
Just as vulnerable in the sense that it is entirely possible, but without the control to mitigate it quickly by revoking all user sessions. The alternative token pattern is basically just the same thing as storing the session in the database, so that seems like a good solution as long as it sufficiently random and un-guessable.

**Comment by joearasin** [comment]
There's no reason for it to be unguessable (It's effectively public, after all). Just need to avoid collisions.

**Comment by simonl2002** [comment]
Reseting alternative token on every logout definitely isn't how I think that was meant to be used. But yeah it effectively does provide server-side control of sessions if we use it.

**Comment by sjohnson540** [comment]
@joearasin are you planning on implementing the alternative token pattern, or would you rather I add sessions to the db, or do we not care about this?

**Comment by joearasin** [comment]
I’ll implement the alternative token thing, but I don’t think logout everywhere when logout is clicked is the proper behavior

**Comment by joearasin** [comment]
And sessions in the DB is probably more trouble than it’s worth

**Comment by sjohnson540** [comment]
Yeah that is why sessions in the database makes sense, you can have multiple sessions open, and logout would invalidate the current session the user was using.

**Comment by sjohnson540** [comment]
It's actually pretty straightforward, there is a extension called Flask-Session that has a Flask-SQLAlchemy adapter that handles it for you.

**Comment by joearasin** [comment]
Does it handle expiration in a reasonable way, or does the DB just accumulate sessions that we need to flush at some point

**Comment by joearasin** [comment]
Mostly, I don’t want to add infrastructure unless it actually solves a problem. I am unconvinced that the additional burden here protects us against a reasonable attack vector

**Comment by jillh510** [comment]
@dryan can this be closed with the merge of #606 ?
